internal/rest/rest_operation.go

```diff
@@ -0,0 +1,248 @@
+		vs.PoolRef = nil
+	}
+	for i := range vs.ServicePoolSelect {
+		//check if normal equal can be made to work
+		if strings.EqualFold(*vs.ServicePoolSelect[i].ServicePoolRef, objRef) {
+			vs.ServicePoolSelect[i].ServicePoolRef = nil
+			utils.AviLog.Infof("Successfully removed Pool ref %s from VS: %s", objRef, *vs.Name)
+		}
+	}
+	op.Obj = vs
+	return true
+}
+
+func (v *virtualserviceSchema) removeVsVipRef(op *utils.RestOp, objRef string) bool {
+	vs, ok := op.Obj.(*avimodels.VirtualService)
+	if !ok {
+		utils.AviLog.Infof("Failed to remove VsVip ref, object is not of type Virtualservice")
+		return false
+	}
+	if strings.EqualFold(*vs.VsvipRef, objRef) {
+		// If VsVip creation failed, then the VS Operations should be aborted
+		utils.AviLog.Infof("VSVip creation failed, object ref won't be removed from Virtualservice")
+		return false
+	}
+	return true
+}
+
+func removeObjRefFromRestOps(restOps []*utils.RestOp, objName, objType string) bool {
+	if _, ok := supportedObjForError[objType]; !ok {
+		utils.AviLog.Debugf("Ignoring error for unsupported type: %v", objType)
+		return false
+	}
+	objRef := "/api/" + objType + "/?name=" + objName
+	for i, op := range restOps {
+		if m, ok := supportedModelTypes[op.Model]; ok {
+			if removeFunc := m.RemoveRef(objType); removeFunc != nil {
+				if !removeFunc(restOps[i], objRef) {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
+func isErrorRetryable(statusCode int) bool {
+	// List of status codes for which we support retry
+	if (statusCode >= 500 && statusCode < 599) || statusCode == 404 || statusCode == 408 || statusCode == 409 {
+		return true
+	}
+	return false
+}
+
+type AviRestClientPool struct {
+	AviClient []*clients.AviClient
+}
+
+func AviRestOperate(c *clients.AviClient, rest_ops []*utils.RestOp) error {
+	for i, op := range rest_ops {
+		SetTenant := session.SetTenant(op.Tenant)
+		SetTenant(c.AviSession)
+		SetVersion := session.SetVersion(op.Version)
+		SetVersion(c.AviSession)
+		switch op.Method {
+		case utils.RestPost:
+			op.Err = c.AviSession.Post(op.Path, op.Obj, &op.Response)
+		case utils.RestPut:
+			op.Err = c.AviSession.Put(op.Path, op.Obj, &op.Response)
+		case utils.RestGet:
+			op.Err = c.AviSession.Get(op.Path, &op.Response)
+		case utils.RestPatch:
+			op.Err = c.AviSession.Patch(op.Path, op.Obj, op.PatchOp,
+				&op.Response)
+		case utils.RestDelete:
+			op.Err = c.AviSession.Delete(op.Path)
+		default:
+			utils.AviLog.Errorf("Unknown RestOp %v", op.Method)
+			op.Err = fmt.Errorf("Unknown RestOp %v", op.Method)
+		}
+		if op.Err != nil {
+			utils.AviLog.Warnf(`RestOp method %v path %v tenant %v Obj %s returned err %s with response %s`,
+				op.Method, op.Path, op.Tenant, utils.Stringify(op.Obj), utils.Stringify(op.Err), utils.Stringify(op.Response))
+			// Wrap the error into a websync error.
+			err := &utils.WebSyncError{Err: op.Err, Operation: string(op.Method)}
+			aviErr := op.Err.(session.AviError)
+			if !isErrorRetryable(aviErr.HttpStatusCode) {
```

Comment:
So looking at this logic, it appears the following might happen:
- if you got a non-retryable error and the operation is not post then you will get rid of the ref from the object model for the first object.
- However, the method would return from Line 241 because of the continue - so if you effectively had more than 1 REST OP that required reference removal, it will not work?

Reply:
Ah looks like it's the exact opposite of what I described that should happen here? That is - you will remove all the references by use of that continue?